ryhope/src/storage/pgsql/storages.rs

```diff
@@ -1320,39 +1523,355 @@ where
+            ensure!(
+                mapper.try_to_incremental_epoch(initial_epoch).await
+                    == Some(INITIAL_INCREMENTAL_EPOCH),
+                "No initial epoch {initial_epoch} found in mapping table {mapper_table_name}"
+            );
+            mapper
+        } else {
+            // add epoch map for `initial_epoch` to the DB
+            db.get()
+                .await?
+                .query(
+                    &format!(
+                        "INSERT INTO {} ({USER_EPOCH}, {INCREMENTAL_EPOCH})
+                            VALUES ($1, $2)",
+                        mapper_table_name,
+                    ),
+                    &[&(initial_epoch as UserEpoch), &INITIAL_INCREMENTAL_EPOCH],
+                )
+                .await?;
+            let cache = InMemoryEpochMapper::new_at(initial_epoch);
+            Self {
+                db,
+                table,
+                in_tx: false,
+                dirty: Default::default(),
+                cache: Arc::new(RwLock::new(cache)),
+            }
+        })
+    }
+
+    /// Add a new epoch mapping for `IncrementalEpoch` `epoch`, assuming that `UserEpoch`s
+    /// are also computed incrementally from an initial shift. If there is already a mapping for
+    /// `IncrementalEpoch` `epoch`, then this function has no side effects, because it is assumed
+    /// that the mapping has already been provided according to another logic.
+    pub(crate) async fn new_incremental_epoch(&mut self, epoch: IncrementalEpoch) -> Result<()> {
+        if let Some(mapped_epoch) = self.cache.write().await.new_incremental_epoch(epoch) {
+            // if a new mapping is actually added to the cache, then we add the `UserEpoch`
+            // of this mapping to the `dirty` set, so that it is later committed to the DB
+            self.dirty.insert(mapped_epoch);
+        }
+        Ok(())
+    }
+
+    pub(crate) fn start_transaction(&mut self) -> Result<()> {
+        ensure!(!self.in_tx, "already in a transaction");
+        self.in_tx = true;
+        Ok(())
+    }
+
+    pub(crate) async fn commit_in_transaction(
+        &mut self,
+        db_tx: &mut Transaction<'_>,
+    ) -> Result<()> {
+        for &user_epoch in self.dirty.iter() {
+            let incremental_epoch = self
+                .cache
+                .write()
+                .await
+                .try_to_incremental_epoch(user_epoch)
+                .await
+                .ok_or(anyhow!("Epoch {user_epoch} not found in cache"))?;
+            db_tx
+                .query(
+                    &format!(
+                        "INSERT INTO {} ({USER_EPOCH}, {INCREMENTAL_EPOCH})
+                     VALUES ($1, $2)",
+                        self.mapper_table_name()
+                    ),
+                    &[&(user_epoch as UserEpoch), &incremental_epoch],
+                )
+                .await?;
```

Comment:
Could we do only _one_ query accross all vs doing one by one for each epoch ?

Reply:
Done in commit https://github.com/Lagrange-Labs/mapreduce-plonky2/pull/425/commits/23affb5f3b0c62fc1ab825043196a2f89e20039c